### PR TITLE
Feature/tao 5939 lock duration authoring

### PR DESCRIPTION
**Pull request by krampstudio**
Related to https://oat-sa.atlassian.net/browse/TAO-5939 

## Time based Item transition - PART I - Authoring

Disclaimer - I had to take shortcuts (especially on the authoring) to wrap this up on time. I'd prefer to rewrite a dedicated component for the timers sync.

This PR brings the following behaviors on the test authoring : 
 On item timeLimits : 
 - if the testPart is linear, the "minimum duration" field is now available, in addition of the maximum duration field.
 - if the testPart is non linear, only the maximum duration is available.
 - if you configure the runtime to support the "guided navigation" : 
`config/taoQtiTest/testRunner.conf.php` : 
```
  'guidedNavigation' => true
```
and the testPart is linear, you have now the possibilty to "lock" the durations in order to enable the "guided navigation mode".

PART II - runtime : https://github.com/oat-sa/extension-tao-testqti/pull/1233

### CODE REVIEW

**Comment by krampstudio** [comment]
The UX strategy has been already approved as well as the use of locked durations. Please share your concerns to @markmolenaar.
For the option, the goal was to enable authoring feature based on the runtime capabilities. I'll move it as a "root" property

File: views/js/controller/creator/creator.js [hunk 1]
```diff
@@ -94,6 +95,7 @@ define([
             options.routes = options.routes || {};
             options.labels = options.labels || {};
             options.categoriesPresets = options.categoriesPresets || {};
+            options.guidedNavigation = options.guidedNavigation === true;
```
[the remark this answers is not in the document]
Comment (krampstudio) [reply views/js/controller/creator/creator.js]: no I just want to ensure the value is actually `true`

**Comment by no-chris** [comment]
@sam please review the UX  as part of the new UX process

**Comment by ssipasseuth** [comment]
I am afraid there is some misunderstandings: could we have a call at 4pm today to discuss it ?

**Comment by ssipasseuth** [comment]
1. the platform level config is "auto-navigation". It should define the behaviour when the item timeouts and it is to be applied when the a max duration is set. It must work independently from the min duration.
2. the lock button is but a helper to set both timers, it could always be there. It should not create any special behaviour but helping setting min=max
3. the min time will just cause the next button to be hidden
4. max duration is better placed before the min duration as it is the more commonly used option (it may be confusing because it has been initially mentioned in the UX channel then we moved to the auto-navigation configuration)
5. I can try proposing some adjustments about the link button in a separate PR
6. would be better to prevent user to set: min duration > max duration

**Comment by ssipasseuth** [comment]
Suggested link icon adaptation here: https://github.com/oat-sa/extension-tao-testqti/pull/1236

**Comment by krampstudio** [comment]
@ssipasseuth 
1.  the selected wording was "guided navigation" Did it changed ?
2. I see 2 issues, the button tooltip has a message dedicated to the guided navigation. If there's no behavior behind, does it make sense to help the user keeping the durations in sync ?
3. Not hidden, disabled for minTime. This is one of the difference between the common mintTime implementation and the guided navigation
4. ok, I'll change it
5. I'll have a look
6. what do you expect when this occurs ? blocking the field values ? warning the user ?

**Comment by ssipasseuth** [comment]
1. "guided navigation" is the macro effect caused by "auto-transition" and max=min, my understanding is that "auto-transition" should be that plateform config.
2. I think that there is no need for help the "guided navigation" as the behaviour is defined 
3. I see ... so following that logic we should have a configuration for the min-time to: 1. disable it, 2. hide it and guided navigation is the result of it ?
6. this is a tricky one, in the item authoring for instance, the min increase pushes the max one (e.g. max choice for a choice interaction) and reducing the max will stops at the min one. But there is an exception because is max=0, the standard consider it infinite (which I agree really does not make sense for a normal innocent user !). I saw a possibility to make the distinction between any value set and disabled. We would need a check in from of the timer. So maybe the easiest for now is to warn the user, what do you think ?

**Comment by krampstudio** [comment]
1. I've checked again, we agreed an "guided navigation" ( from @markmolenaar ) but I can change all mentions to auto transitions.
2. This is the same, we agreed to add a tooltip on the link button. I can remove it as well.
3. by default minTime disables the navigation, this is the QTI defined behavior. So if you define a minTime the navigation will be disabled until the countdown goes over. The hiding them is a pure artifact of the custom behavior "guided navigation"
4. I'll see what I can do.

**Comment by ssipasseuth** [comment]
4. looks good for a quick and dirty solution... I have been thinking on my way home. In the future, if we want to change the pattern, we could try something like that:
min-max-duration

**Comment by ssipasseuth** [comment]
2. Sorry, you are right, the help tooltip is very useful to make him aware that he is in a special case.
1. I think I finally understand the way you implemented the solution. There were multiple discussions around this story and each discussion focused on different aspect of it which is causing a lot of confusions. The current implementation is suggesting that the link button will activate the guided-navigation and implicitly the auto-transition.
The issue is only on the last bit about “implicit auto-transition” indeed we decided to add the platform “auto-navigation” config  to prevent the implicit (max=min) => auto-navigation and to support additional customers’ requirements.
So my understanding is the platform config is “auto-navigation” and it configures behaviour on max duration only. 
That is a big difference for the user as the user having the guidedNavigation on his instance now will have 2 types of behaviour on max timeout, when max=min and max!=min.
What I would expect instead is that the max time always has the same behaviour either auto-navigate or not.

The definition I will give to each concept:
guided navigation=(when min=max) you cannot decided when you can make the transition, but it does not define what happens before and after it. 
auto-transition=(platform config) move to next item when max time is reached (without user dialog)
hide-next=(probably needed platform config, works well with “auto-transition") hide the next button until the min time is reached
That is why I thought: there is no reason why the guided navigation and its link button are not always available. And with the available platform config (and later item category) you can customize the experience of your guided navigation: you may want to deactivate auto-transition by design because you want to let the test taker rest and get ready for the next item for instance. So we could decide to also display a message between and ask the user to click on a button when he is ready to go to the next item.

But having those strategy in place is definitely a good thing because it will allows us to implement the right behaviour.
All I am saying is that it is recommended to have no magic combination and one user option = one single deterministic behaviour (unless it is a randomise action). This is important to let the user feel in control of what they do.

I think we have some common understanding, we will just need to merge them together :)

**Pull request merged**